File: ts/src/actions/Execution.ts
import {constructValue, PlainValue} from "@code0-tech/tucana/helpers";
import {ActionExecutionRequest, ActionExecutionResponse, ActionTransferRequest} from "@code0-tech/tucana/aquila";
import {NodeExecutionResult, Error as ProtoError} from "@code0-tech/tucana/shared";
import {FunctionContext, RuntimeError} from "../types";
import {RuntimeFunctionProps} from "../models/runtime_function.model";
import {ResolvedValue, resolveNodeValue} from "../internal/literal";
import {CodeZeroEvent} from "../events";
import type {Action} from "../action";

export const packetType = "execution";

function nowMicros(): bigint {
    return BigInt(Math.floor((performance.timeOrigin + performance.now()) * 1000));
}

function buildParams(action: Action, execution: ActionExecutionRequest, func: RuntimeFunctionProps): (ResolvedValue | undefined)[] {
    return (func.parameters || []).map((_param, index) =>
        resolveNodeValue(action, execution.parameters?.[index]),
    );
}

export function handle(action: Action, execution: ActionExecutionRequest): void {
    action.emit(CodeZeroEvent.executionRequestReceived, execution);

    const func = action.runtimeFunctions.get(execution.functionIdentifier);
    if (!func) {
        console.error("Received execution request for unknown function:", execution.functionIdentifier);
        return;
    }

    const params = buildParams(action, execution, func);

    const conf = action.configs.get(execution.projectId) ?? {
        projectId: 0n,
        configValues: [],
        findConfig: () => undefined,
    };

    const context: FunctionContext = {
        projectId: execution.projectId,
        executionId: execution.executionIdentifier,
        matchedConfig: conf,
        executeFlow: (flowId, payload) => action.executeFlow(flowId, payload),
    };

    const startedAt = nowMicros();
    const funcHandler = func.handler;
    const allParams: (PlainValue | undefined)[] =
        funcHandler.length === params.length + 1 ? [context as PlainValue, ...params] : params;

    Promise.resolve(funcHandler(...allParams)).then((value: PlainValue) => {
        const finishedAt = nowMicros();
        action.stream!.requests.send(ActionTransferRequest.create({
            data: {
                oneofKind: "result",
                result: ActionExecutionResponse.create({
                    executionIdentifier: execution.executionIdentifier,
                    nodeResult: NodeExecutionResult.create({
                        id: {oneofKind: undefined},
                        startedAt,
                        finishedAt,
                        parameterResults: [],
                        result: {oneofKind: "success", success: constructValue(value ?? null)},
                    }),
                }),
            },
        })).catch(err => console.error("Failed to send execution result:", err));

    }).catch((error: unknown) => {
        const finishedAt = nowMicros();
        const isRuntimeError = error instanceof RuntimeError;
        action.stream!.requests.send(ActionTransferRequest.create({
            data: {
                oneofKind: "result",
                result: ActionExecutionResponse.create({
                    executionIdentifier: execution.executionIdentifier,
                    nodeResult: NodeExecutionResult.create({
                        id: {oneofKind: undefined},
                        startedAt,
                        finishedAt,
                        parameterResults: [],
                        result: {
                            oneofKind: "error",
                            error: ProtoError.create({
                                code: isRuntimeError ? (error as RuntimeError).code : "UNKNOWN_ERROR",
                                category: "RUNTIME",
                                message: isRuntimeError
                                    ? ((error as RuntimeError).description || (error as Error).message)
                                    : String(error),
                                timestamp: finishedAt,
                                version: action.version,
                                dependencies: {},
                            }),
                        },
                    }),
                }),
            },
        })).catch(err => console.error("Failed to send execution error:", err));
    });
}

File: py/hercules/actions/execution.py
"""Execution request handler (port of ``src/actions/Execution.ts``)."""
from __future__ import annotations

import asyncio
import inspect
import time

from hercules._tucana.helpers import construct_value, to_allowed_value
from tucana.generated.aquila import action_pb2
from tucana.generated.shared import errors_pb2, execution_result_pb2
from hercules.events import CodeZeroEvent
from hercules.types import ProjectConfiguration, RuntimeError

packet_type = "execution"


def _now_micros() -> int:
    return time.time_ns() // 1000


def _build_params(action, execution, func):
    params = []
    parameters = func.parameters or []
    for index, _param in enumerate(parameters):
        field = execution.parameters[index] if index < len(execution.parameters) else None
        which = field.WhichOneof("value") if field is not None else None
        if which == "literal_value":
            params.append(to_allowed_value(field.literal_value.value))
        elif which == "sub_flow":
            sub_flow = field.sub_flow

            def make_caller(sub_flow):
                async def caller(*args):
                    return await action.execute_sub_flow(sub_flow, *args)

                return caller

            params.append(make_caller(sub_flow))
        else:
            params.append(None)
    return params


def _wants_context(handler, param_count: int) -> bool:
    try:
        sig = inspect.signature(handler)
    except (TypeError, ValueError):
        return False
    positional = [
        p
        for p in sig.parameters.values()
        if p.kind in (p.POSITIONAL_ONLY, p.POSITIONAL_OR_KEYWORD)
    ]
    # A `*args` handler has no fixed positionals and never gets a context.
    return len(positional) == param_count + 1


def handle(action, execution) -> None:
    action.emit(CodeZeroEvent.execution_request_received, execution)

    func = action.runtime_functions.get(execution.function_identifier)
    if func is None:
        print("Received execution request for unknown function:", execution.function_identifier)
        return

    params = _build_params(action, execution, func)

    conf = action.configs.get(execution.project_id) or ProjectConfiguration(
        project_id=0,
        config_values=[],
        find_config=lambda identifier: None,
    )

    from hercules.types import FunctionContext

    context = FunctionContext(
        project_id=execution.project_id,
        execution_id=execution.execution_identifier,
        matched_config=conf,
        execute_flow=lambda flow_id, payload=None: action.execute_flow(flow_id, payload),
    )

    started_at = _now_micros()
    handler = func.handler
    all_params = (
        [context, *params] if _wants_context(handler, len(params)) else params
    )

    asyncio.ensure_future(_run(action, execution, handler, all_params, started_at))


async def _run(action, execution, handler, all_params, started_at) -> None:
    try:
        value = handler(*all_params)
        if inspect.isawaitable(value):
            value = await value
        finished_at = _now_micros()
        request = action_pb2.ActionTransferRequest(
            result=action_pb2.ActionExecutionResponse(
                execution_identifier=execution.execution_identifier,
                node_result=execution_result_pb2.NodeExecutionResult(
                    started_at=started_at,
                    finished_at=finished_at,
                    parameter_results=[],
                    success=construct_value(value if value is not None else None),
                ),
            )
        )
        await action.send(request)
    except Exception as error:  # noqa: BLE001
        finished_at = _now_micros()
        is_runtime_error = isinstance(error, RuntimeError)
        proto_error = errors_pb2.Error(
            code=error.code if is_runtime_error else "UNKNOWN_ERROR",
            category="RUNTIME",
            message=(error.description or str(error)) if is_runtime_error else str(error),
            timestamp=finished_at,
            version=action.version,
        )
        request = action_pb2.ActionTransferRequest(
            result=action_pb2.ActionExecutionResponse(
                execution_identifier=execution.execution_identifier,
                node_result=execution_result_pb2.NodeExecutionResult(
                    started_at=started_at,
                    finished_at=finished_at,
                    parameter_results=[],
                    error=proto_error,
                ),
            )
        )
        try:
            await action.send(request)
        except Exception as err:  # noqa: BLE001
            print("Failed to send execution error:", err)
